#1903: TTIR Tile Compute Ops

jdesousa-TT opened the pull request:
Ticket #1903 

Adds:

- `ttir.tile_add(tt.tile, tt.tile)`
- `ttir.tile_sub(tt.tile, tt.tile)`
- `ttir.tile_mul(tt.tile, tt.tile)`
- `ttir.tile_recip(tt.tile)`
- `ttir.tile_exp(tt.tile)`
- `ttir.tile_log(tt.tile)`
- `ttir.tile_reduce_sum(tt.tile, tt.tile)`
- `ttir.tile_reduce_max(tt.tile)`
- `ttir.tile_matmul_block(memref<tt.tile>, memref<tt.tile>)` This one is a special case that operates on a block.

Adds a TTIR_GenericParent interface that verifies these tile ops have a TTIR GenericOp as their ancestor (with the exception of ttir.yield because of its use in TTNN dialect).

TODO: 

- [x] Determine if these tile ops should have result types @nsmithtt ?
- [x] Add verifier for `matmul_block` inputs. i.e. ensure the memref refers to tiles
- [x] Brief tests

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
```diff
@@ -0,0 +1,99 @@
+include "ttmlir/Dialect/TTIR/IR/TTIRBase.td"
+include "ttmlir/Dialect/TTIR/IR/TTIROpsAttrs.td"
+include "ttmlir/Dialect/TTIR/IR/TTIROpsInterfaces.td"
+
```
It might be nice to have:
```td
class TTIR_TileOp<string mnemonic, list<Trait> traits = []> :
    TTIR_Op<mnemonic, !listconcat([TTIR_GenericParent], traits)> {
} 
```

Then:
```td
def TTIR_TileAddOp : TTIR_TileOp<"tile_add"> {
```

jdesousa-TT replied:
Done

nsmithtt commented on lib/Dialect/TTIR/IR/TTIROpsInterfaces.cpp:
```diff
@@ -47,3 +48,18 @@ mlir::tt::ttir::detail::verifyBroadcastable(mlir::Operation *op) {
+    return op->emitOpError("TTIR Tile Ops must be inside a generic region");
+  }
+
+  if (llvm::dyn_cast_or_null<ttir::GenericOp>(parent)) {
```
I think `dyn_cast_or_null`, which forwards to `dyn_cast_if_present`, will return null if `parent` is null.  In other words, I think the check on line 56 is redundant and we can probably safely remove.

jdesousa-TT replied:
Thanks, changed.

nsmithtt reviewed (commented):
@jdesousa-TT, this looks great!  Minor comments inline

vroubtsovTT commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
```diff
@@ -0,0 +1,99 @@
+def TTIR_TileReduceSumOp : TTIR_Op<"tile_reduce_sum", [TTIR_GenericParent]>{
+    let summary = "TTIR Tile Reduce Sum Op";
+    let description = [{
+        The `tile_reduce_sum` operation computes the sum of all elements in the input tensor.
```
note that metal reduce ops have more going on:

1. there is a way to specify which dims get reduced, rows, cols, or both
2. the reduction op can be "sum" or "max"

jdesousa-TT replied:
I think the overarching metal reduce ops have more going on, which is more synonymous to ttir.sum/ttir.max, etc. 

but for the llk reduce_tile, which this should be synonymous with, the only inputs are two tiles + reduce dim:

```cpp
template <PoolType reduce_type = REDUCE_OP, ReduceDim reduce_dim = REDUCE_DIM>
ALWI void reduce_tile(uint32_t icb0, uint32_t icb1, uint32_t itile0, uint32_t itile1, uint32_t idst) {
    MATH((llk_math_reduce<reduce_type, reduce_dim, MATH_FIDELITY, DST_ACCUM_MODE>(idst)));
    UNPACK((llk_unpack_AB(icb0, icb1, itile0, itile1)));
}
```

We should add a `ttir.tile_reduce_max` to differentiate from sum, and possibly an attribute on both for the reduce dim.

vroubtsovTT replied:
Yes, the llk reduce_tile is what I meant. "sum-reduce" does not imply "sum of all elements", it should be parameterized.

nsmithtt commented on lib/Dialect/TTIR/IR/TTIROpsInterfaces.cpp:
```diff
@@ -47,3 +49,18 @@ mlir::tt::ttir::detail::verifyBroadcastable(mlir::Operation *op) {
+
+  while (parent) {
+    if (llvm::dyn_cast<ttir::GenericOp>(parent) ||
+        llvm::dyn_cast<linalg::GenericOp>(parent)) {
```
I think we want to forego the `linalg` check because it too must also appear inside of a ttir.generic

nsmithtt commented:
Have folks given this anymore thought? Are we OK to continue review & land?

azecevicTT commented on lib/Dialect/TTIR/IR/TTIROps.cpp:
If we are going with this solution we should also split lib file to `lib/Dialect/TTIR/IR/TTIRTileOps.cpp`. These two classes of ops are largely independent from each other, so we should at least try to split them through file organization as much as possible.

nsmithtt replied:
Agreed

azecevicTT commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
```diff
@@ -0,0 +1,103 @@
+include "ttmlir/Dialect/TTIR/IR/TTIROpsAttrs.td"
+include "ttmlir/Dialect/TTIR/IR/TTIROpsInterfaces.td"
+
+class TTIR_TileOp<string mnemonic, list<Trait> traits = []> :
```
Due to a very different nature of tile and non-tile ops, I can see that working with some unknown TTIR op we might want a different behaviour depending if it's tile op or not. As far as I know, tablegen classes don't translate in any way to some C++ type information, so there isn't some easy way to obtain that info. One way would be to try to match the name as a string, but that's really ugly and unsafe way to do it. The other one that I can think of would be to introduce some simple trait that would act like a label, something like `TileOpTrait` that can be attached to `TTIR_TileOp`, so we can do something like this
```cpp
if (op->hasTrait<TileOpTrait>()) {
  // do something
} else {
  // do something else
}
```

nsmithtt replied:
Yeah that's a good callout, @jdesousa-TT let's add a tile trait

azecevicTT reviewed (commented):
I can see a pros and cons of both approaches (this and the new dialect), `TTIR_GenericOp` definitely being a strong point for this approach.

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
```diff
@@ -0,0 +1,103 @@
+    }];
+
+    let arguments = (ins TT_Tile:$lhs, TT_Tile:$rhs);
+    let results = (outs TT_Tile:$result); // questionable: should tile ops actually return anything? tensix math ops will leave results in dest regs, up to packer to get result to l1
```
@jdesousa-TT, yes these ops should return a tile result.  At this level the values will just be virtual registers and register allocation will happen later.

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
```diff
@@ -0,0 +1,103 @@
+        The `tile_reduce_sum` operation computes the sum of all elements in the input tensor.
+    }];
+
+    let arguments = (ins TT_Tile:$a, TT_Tile:$b);
```
Yes, @jdesousa-TT let's add that to the description that would be useful.

nsmithtt commented on lib/Dialect/TTIR/IR/TTIROpsInterfaces.cpp:
```diff
@@ -47,3 +49,18 @@ mlir::tt::ttir::detail::verifyBroadcastable(mlir::Operation *op) {
+
+mlir::LogicalResult
+mlir::tt::ttir::detail::verifyGenericParent(mlir::Operation *op) {
+  mlir::Operation *parent = op->getParentOp();
```
Good tip!

azecevicTT replied:
Consider this once more, it's implemented using the same loop structure https://mlir.llvm.org/doxygen/IR_2Operation_8h_source.html#l00238.

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIROpsInterfaces.td:
```diff
@@ -124,4 +124,12 @@ def TTIR_FullyBroadcastable : OpInterface<"FullyBroadcastable"> {
+  let verify = [{
+    return detail::verifyGenericParent($_op);
+  }];
+}
```
Instead of this I found `class HasParent<string op>` inside of `OpBase.td`.  I think we can just write:

```c++
class TTIR_TileOp<string mnemonic, list<Trait> traits = []> :
    TTIR_Op<mnemonic, !listconcat([HasParent<"GenericOp">], traits)> {
```

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
I wonder if we actually want to call this `TTIRGenericRegionOps`, then we can put `YieldOp` here too, plus we might want some data movement related ones in the future...

nsmithtt replied:
I think we should go ahead with this rename and move YieldOp there too.  Go ahead and leave `TTIR_KernelOp` where it is, we should be able to completely remove it after the d2m refactor.

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIROpsInterfaces.h:
```diff
@@ -13,6 +13,7 @@ namespace tt {
 namespace ttir {
 namespace detail {
 mlir::LogicalResult verifyBroadcastable(mlir::Operation *op);
+mlir::LogicalResult verifyGenericParent(mlir::Operation *op);
```
Is this still needed?

azecevicTT commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
```diff
@@ -0,0 +1,121 @@
+
+    let arguments = (ins TT_Tile:$a, 
+                         TT_Tile:$b,
+                         OptionalAttr<I32ArrayAttr>:$dim_arg);
```
Is `dim_arg` supposed to represent dimensions we are reducing over? I'm not sure this is the best way to do that. `OptionalAttr` means you can get away without writing in the IR. What's that supposed to mean, that we are reducing over RC? What would values in the array represent in that case? You would need to write a verifier in this case.

The sound alternatives that I see are a proper EnumAttr that would emulate the `ReduceDim` from kernel with `R`, `C` and `RC` value.
The second alternative would be `OptionalAttr<I1Attr>`, with a semantics of `0 -> R`, `1 -> C`, `lack of existence -> RC`.

First option would require writing a custom attribute, but it's the safest and the most explicit option.

vroubtsovTT replied:
Note that we already have a group of reduction ops in `TTIROps`, also with a base/derived structure:

```
class TTIR_ReductionOp<string mnemonic, list<Trait> traits = []> :
    TTIR_DPSOp<mnemonic, !listconcat([TTIR_GenericRegionOpInterface], traits)> {

    let summary = "Reduction op.";
    let description = [{
      Reduction op.
    }];

    let arguments = (ins AnyRankedTensor:$input,
                         AnyRankedTensor:$output,
                         BoolAttr:$keep_dim,
                         OptionalAttr<I32ArrayAttr>:$dim_arg);

    let results = (outs AnyRankedTensor:$result);
...
def TTIR_SumOp : TTIR_ReductionOp<"sum"> {
    let summary = "Sum reduction op.";
    let description = [{
      Sum reduction op.
    }];
}

def TTIR_MeanOp : TTIR_ReductionOp<"mean"> {
  let summary = "Mean reduction op.";
  let description = [{
    Mean reduction op.
  }];
}

def TTIR_MaxOp : TTIR_ReductionOp<"max"> {
  let summary = "Max reduction op.";
  let description = [{
    Max reduction op.
  }];
}
```
Do we plan to remove them? If now, we probably want to preserve a similar inheritance structure for the `ttir.tile_reduce_...` group of ops and have similar signatures across both groups?

Something feels unsatisfying about deaIing with a different operand type with a new op name prefix. I wish tablegen supported overloading... In fact, is there a way to fake it via something like
```
    let arguments = (ins AnyTypeOf<[AnyRankedTensor, TT_Tile]>:$input,
                         AnyTypeOf<[AnyRankedTensor, TT_Tile]>:$output,
                         BoolAttr:$keep_dim,
                         OptionalAttr<I32ArrayAttr>:$dim_arg);

```
after introducing a new operand type `TT_Tile`?

azecevicTT replied:
This would be really stretch and could easily lead to some weird bugs where `input` is `AnyRankedTensor` and `output` is `TT_Tile`. Also, the fact that both take two tensors/tiles is almost accidental, because one is an actual input (in the case of tile), and the other is DPS output.

We could have something like `class TTIR_TileReductionOps`. The thing with tablegen classes is that they can help us organize a code a bit better, but we shouldn't rely too much on the intuition that we have from OOP because they lack many properties of OOP classes (they don't translate to cpp classes themselves).

vroubtsovTT replied:
> This would be really stretch and could easily lead to some weird bugs where `input` is `AnyRankedTensor` and `output` is `TT_Tile`.

If not easily done via tablegen syntax, this constraint could be enforced via `verify()` created in response to `let genVerifyDecl = 1;` in the `def`.

MLIR has the same design problem in some existing dialects, e.g. quoting `LinalgStructuredOps.td`:
```
def TensorOrMemref :
  AnyTypeOf<[AnyMemRef, AnyRankedTensor], "", "::mlir::ShapedType">;
...
def ReduceOp : LinalgStructuredBase_Op<"reduce", [
    ..., SameVariadicOperandSize, ...]> {
  let summary = "Reduce operator";
  let description = [{
    Executes `combiner` on the `dimensions` of `inputs` and returns the
    reduced result. The `dimensions` attribute needs to list the reduction
    dimensions in increasing order.

    Example:
    ```
      %reduce = linalg.reduce
          ins(%input:tensor<16x32x64xf32>)
          outs(%init:tensor<16x64xf32>)
          dimensions = [1]
          (%in: f32, %out: f32) {
            %0 = arith.addf %out, %in: f32
            linalg.yield %0: f32
          }
    ```
...
  }];

  let arguments = (ins
    Variadic<TensorOrMemref>:$inputs,
    Variadic<TensorOrMemref>:$inits,

    ConfinedAttr<DenseI64ArrayAttr,
                 [DenseArrayStrictlySorted<DenseI64ArrayAttr>]>:$dimensions
  );
  let results = (outs Variadic<AnyTensor>);
...
```
Our need is actually very similar to linalg.generic's and these ops will participate in linalg.generic-like lowerings. Notice how their `ReduceOp` takes a variadic list of `AnyTypeOf<[AnyMemRef, AnyRankedTensor]`. They enforce `SameVariadicOperandSize` as a constraint trait and MLIR also has `SameTypeOperands`  and `SameOperandsAndResultType`. And if those can't be made to work, there is always the c++ impl we can put into `verify()`.

> but we shouldn't rely too much on the intuition that we have from OOP because they lack many properties of OOP classes (they don't translate to cpp classes themselves).

You are correct. We understand that this is not going to translate into C++ OOP. (C++ OOP is not very useful to us anyway because of how `llvm::dyn_cast<> & Co` work anyway.) Inheriting from a common tablegen `class` will help inheriting some structure like a MLIR Trait, Interface, etc.

We are looking for any design hooks we can put in that can help us match against these ops as a group when writing MLIR type converters/rewriters/etc without having to always list all of the ops everywhere. There's probably more we can borrow from how linalg deals with this. On the positive side, these can probably be added later; what I suggesting to give some thought to right now is whether we're really ok with adding to our frontend IR a whole group of ops that are like an existing group except for their operand types (that only materialize much later, close to the metal).

azecevicTT replied:
I generally see a value in such generic form, but in this concrete case, these reduce ops differs significantly depending on whether they operate on a tensor or a tile, the biggest difference is that tile ops are weighted as I understand it. Take in mind that I'm not totally familiar with the whole process of lowering (do have some familiarity with `linalg.generic` though), so would love to hear it from someone more familiar with it how would we approach this problem.

vroubtsovTT replied:
If everybody is fine with adding these as a distinct "backend" group of ops to our frontend IR, I retract my objections ;)

nsmithtt replied:
We should definitely have `tile_reduce_sum` & `tile_reduce_max` as new "tile" ops, they belong to a different semantic class than their tensor based relatives.  And I agree with @azecevicTT:
- Dim should not be optional
- Let's introduce an enum in `TTIROpsEnums.td`, `ReduceDim { R, C, RC };`

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIRTileOps.td:
```diff
@@ -0,0 +1,121 @@
+def TTIR_TileReduceSumOp : TTIR_TileOp<"tile_reduce_sum">{
+    let summary = "TTIR Tile Reduce Sum Op";
+    let description = [{
+        The `tile_reduce_sum` operation computes the sum of all elements in the input tensor over the specified reduction dims.
```
+1

jdesousa-TT commented:
Hey @nsmithtt @vroubtsovTT @azecevicTT,

I've addressed all of the feedback from this PR, if you could re-review and let me know that would be great.

azecevicTT commented on include/ttmlir/Dialect/TTIR/IR/TTIRBase.td:
```diff
@@ -65,5 +65,7 @@ def TTIR_Idempotence : TTIR_Trait<"TTIRIdempotence", [DestinationStyleOpInterfac
 // Example: and(x, x) == x
 def TTIR_BinaryIdempotence : TTIR_Trait<"TTIRBinaryIdempotence", [DestinationStyleOpInterface, ThreeOperands, NoMemoryEffect]>;
+// GenericRegionOpTrait is a trait that acts as a label for all generic region operations.
+def TTIR_GenericRegionOpTrait : TTIR_Trait<"TTIRGenericRegionOpTrait", []>;
```
I'm wondering if we should define `TTIR_GenericRegionOpTrait` as `def TTIR_GenericRegionOpTrait : HasParent<"GenericOp">;`? Also will `GenericOp` be ambiguous, should we prepend a namespace? @nsmithtt

jdesousa-TT replied:
One thing to note with this is we're forced to include `ScatterOp` as one of the valid parents for `YieldOp` because it is used in TTNN Scatter.

Also, the HasParent trait only checks the immediate parent, meaning that we should either go back to the custom verifier that checks recursively, or enforce that the tile ops are only valid if their direct parent is a linalg generic.

I'm not sure if enforcing linalg generic will work if there's a step in the lowering that the tile op exists before we add the linalg generic region.

Thoughts? @nsmithtt @azecevicTT

nsmithtt replied:
@jdesousa-TT, good catch, `HasParent` does fail for me locally when these ops end up being nested inside of linalg or scf control flow.  I think we have to go back to the recursive approach, thank you for trying it!

jdesousa-TT replied:
Updated this to match a ttir:GenericOp specifically via a manual verifier. This allows us to check recursively.

I've left `ttir.yield` out of the verifier scope because there are some uses of it outside of `ttir.generic` in the TTNN path.

azecevicTT commented on include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.h:
You should include this into `lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp` (it won't compile without it).

jdesousa-TT replied:
Good catch, ty

azecevicTT reviewed (commented):
This looks good now, with a few points mentioned.

You should start with writing some tests, there isn't much to test yet, but you can still test the structure (ops inside `GenericOp`), some attributes that you've written and stuff like that.

This is a good tutorial if you are new to FileCheck tests https://llvm.org/docs/CommandGuide/FileCheck.html#tutorial.

jdesousa-TT commented on include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td:
```diff
@@ -0,0 +1,133 @@
+        The `tile_reduce_max` operation computes the max of all elements in the input tile over the specified reduction dim(s).
+    }];
+
+    let arguments = (ins TT_Tile:$input,
```
@nsmithtt Can you confirm this? I'm struggling to see how reduce_max will take two tile inputs. But I agree that the metal llk does indeed take two input tiles for sum & max.

nsmithtt replied:
Wow, I suppose @azecevicTT is right, I thought there was a separate API.  What could a scalar possible mean or do for reduce max? 😭

@xanderchin, do you know if GAPool or GMPool expects something in both src registers?  Or is this just a quirk of the llk op?

xanderchin replied:
GAPool requires 1/(total elems) in src reg for the coefficient. 

I think MaxPool requires some constants though I don't recall the spec exactly.

jdesousa-TT commented:
Added some tests to make sure the ops are valid, the reduce_dim attribute is working (and being checked), and that the generic parent verifier is working.

vroubtsovTT commented on lib/Dialect/TTIR/IR/TTIROpsInterfaces.cpp:
```diff
@@ -47,3 +48,16 @@ mlir::tt::ttir::detail::verifyBroadcastable(mlir::Operation *op) {
+  mlir::Operation *parent = op->getParentOp();
+  while (parent) {
+    if (llvm::dyn_cast_or_null<ttir::GenericOp>(parent)) {
+      return success();
```
A small nit: since the `while ensures` that parent is not null inside the loop, using `llvm::dyn_cast_or_null` instead of `llvm::dyn_cast` seems redundant?

jdesousa-TT replied:
I suppose this is better, changed :)

jdesousa-TT commented on test/ttmlir/Dialect/TTIR/generic/generic_region_ops_negative.mlir:
```diff
@@ -0,0 +1,39 @@
+
+// -----
+
+func.func @not_in_generic_region(%a: !tt.tile<32x32, f32>, %b: !tt.tile<32x32, f32>) -> !tt.tile<32x32, f32> {
```
Fixed this, I accidentally pasted over the correct test criteria :(

This is supposed to test the generic region verifier.

nsmithtt commented on include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td:
```diff
@@ -0,0 +1,134 @@
+
+    let arguments = (ins AnyNon0RankedMemRef:$a,
+                         AnyNon0RankedMemRef:$b);
+    let results = (outs AnyNon0RankedMemRef:$result);
```
Nit pick on the matmul block, we should make it DPS, it's otherwise impossible to use since we're given the output memref as argument to the region body.

```
def TTIR_TileMatmulBlockOp : TTIR_TileOp<"tile_matmul_block",    
  [ DestinationStyleOpInterface    
  , MemoryEffects<[MemRead, MemWrite]>    
  ]> {    
    let summary = "TTIR Tile Matmul Block Op";    
    let description = [{    
        The `tile_matmul_block` operation computes the matrix multiplication of two input blocks.    
    }];    
    
    let arguments = (ins AnyNon0RankedMemRef:$a,    
                         AnyNon0RankedMemRef:$b,    
                         AnyNon0RankedMemRef:$output);    
    
    let extraClassDeclaration = [{    
      MutableOperandRange getDpsInitsMutable() { return getOutputMutable(); }    
    }];    
    
    let hasVerifier = 1;    
}
```

jdesousa-TT replied:
Good callout, changed

azecevicTT reviewed (approved):
Looks good, thanks for writing some tests.
